src/common/task.cc

```diff
@@ -348,7 +359,93 @@ const std::unordered_map<std::string, double> TaskSpec_get_required_resources(
+  TaskSpec *copy = (TaskSpec *) malloc(task_spec_size);
+  memcpy(copy, spec, task_spec_size);
+  return copy;
+}
+
+void TaskSpec_free(TaskSpec *spec) {
+  free(spec);
+}
+
+TaskExecutionSpec::TaskExecutionSpec(
+    const std::vector<ObjectID> &execution_dependencies,
+    TaskSpec *spec,
+    int64_t task_spec_size) {
+  execution_dependencies_ = execution_dependencies;
+  task_spec_size_ = task_spec_size;
+  TaskSpec *spec_copy = new TaskSpec[task_spec_size_];
+  memcpy(spec_copy, spec, task_spec_size);
+  spec_ = std::unique_ptr<TaskSpec[]>(spec_copy);
+}
+
+TaskExecutionSpec::TaskExecutionSpec(TaskExecutionSpec *other) {
+  execution_dependencies_ = other->execution_dependencies_;
+  task_spec_size_ = other->task_spec_size_;
+  TaskSpec *spec_copy = new TaskSpec[task_spec_size_];
+  memcpy(spec_copy, other->spec_.get(), task_spec_size_);
+  spec_ = std::unique_ptr<TaskSpec[]>(spec_copy);
+}
+
+std::vector<ObjectID> TaskExecutionSpec::ExecutionDependencies() {
+  return execution_dependencies_;
+}
+
+int64_t TaskExecutionSpec::SpecSize() {
+  return task_spec_size_;
+}
+
+TaskSpec *TaskExecutionSpec::Spec() {
+  return spec_.get();
+}
+
+int64_t TaskExecutionSpec::NumDependencies() {
+  TaskSpec *spec = Spec();
+  int64_t num_dependencies = TaskSpec_num_args(spec);
+  num_dependencies += execution_dependencies_.size();
+  return num_dependencies;
+}
+
+int TaskExecutionSpec::DependencyIdCount(int64_t dependency_index) {
+  TaskSpec *spec = Spec();
+  /* The first dependencies are the arguments of the task itself, followed by
+   * the execution dependencies. Find the total number of task arguments so
+   * that we can index into the correct list. */
+  int64_t num_args = TaskSpec_num_args(spec);
+  if (dependency_index < num_args) {
+    /* Index into the task arguments. */
+    return TaskSpec_arg_id_count(spec, dependency_index);
+  } else {
+    /* Index into the execution dependencies. */
+    dependency_index -= num_args;
+    CHECK((size_t) dependency_index < execution_dependencies_.size());
+    /* All elements in the execution dependency list have exactly one ID. */
+    return 1;
+  }
+}
+
+ObjectID TaskExecutionSpec::DependencyId(int64_t dependency_index,
+                                         int64_t id_index) {
+  TaskSpec *spec = Spec();
+  /* The first dependencies are the arguments of the task itself, followed by
+   * the execution dependencies. Find the total number of task arguments so
+   * that we can index into the correct list. */
+  int64_t num_args = TaskSpec_num_args(spec);
+  if (dependency_index < num_args) {
+    /* Index into the task arguments. */
+    return TaskSpec_arg_id(spec, dependency_index, id_index);
+  } else {
+    /* Index into the execution dependencies. */
+    dependency_index -= num_args;
+    CHECK((size_t) dependency_index < execution_dependencies_.size());
+    return execution_dependencies_[dependency_index];
+  }
+}
+
+bool TaskExecutionSpec::DependsOn(ObjectID object_id) {
```

Comment:
another instance where we should really be passing a const ref to the object id. Does a few useful things:
* establishes a contract with the caller that the object_id is not going to be mutated
* eliminates a 20 byte copy when calling this function